src/connection.cpp

```diff
@@ -41,295 +41,306 @@ namespace oxen::quic
-            log::trace(log_cat, "HANDSHAKE CONFIRMED on CLIENT connection");
+            rv = conn->server_handshake_completed();
 
             if (conn->conn_established_cb)
                 conn->conn_established_cb(*conn);
             else
                 conn->endpoint().connection_established(*conn);
-
-            return 0;
         }
+        else
+            rv = conn->client_handshake_completed();
 
-        static void rand_cb(uint8_t* dest, size_t destlen, const ngtcp2_rand_ctx* rand_ctx)
-        {
-            (void)rand_ctx;
-            (void)gnutls_rnd(GNUTLS_RND_RANDOM, dest, destlen);
-        }
+        return rv;
+    }
 
-        static int on_connection_id_status(
-                ngtcp2_conn* /* _conn */,
-                ngtcp2_connection_id_status_type type,
-                uint64_t /* seq */,
-                const ngtcp2_cid* cid,
-                const uint8_t* /* token */,
-                void* user_data)
-        {
-            auto* conn = static_cast<Connection*>(user_data);
+    int connection_callbacks::on_handshake_confirmed(ngtcp2_conn*, void* user_data)
+    {
+        auto* conn = static_cast<Connection*>(user_data);
 
-            auto dir_str = conn->is_inbound() ? "SERVER"s : "CLIENT"s;
-            auto action = type == NGTCP2_CONNECTION_ID_STATUS_TYPE_ACTIVATE ? "ACTIVATING"s : "DEACTIVATING"s;
-            log::trace(log_cat, "{} {} DCID:{}", dir_str, action, oxenc::to_hex(cid->data, cid->data + cid->datalen));
+        // server should never call this, as it "confirms" on handshake completed
+        assert(conn->is_outbound());
+        log::trace(log_cat, "HANDSHAKE CONFIRMED on CLIENT connection");
 
-            // auto& ep = conn->endpoint();
+        if (conn->conn_established_cb)
+            conn->conn_established_cb(*conn);
+        else
+            conn->endpoint().connection_established(*conn);
 
-            switch (type)
-            {
-                case NGTCP2_CONNECTION_ID_STATUS_TYPE_ACTIVATE:
-                    // ep.associate_cid(cid, *conn);
-                    break;
-                case NGTCP2_CONNECTION_ID_STATUS_TYPE_DEACTIVATE:
-                    // ep.dissociate_cid(cid, *conn);
-                    break;
-                default:
-                    break;
-            }
+        return 0;
+    }
 
-            return 0;
-        }
+    void connection_callbacks::rand_cb(uint8_t* dest, size_t destlen, const ngtcp2_rand_ctx* rand_ctx)
+    {
+        (void)rand_ctx;
+        (void)gnutls_rnd(GNUTLS_RND_RANDOM, dest, destlen);
+    }
+
+    int connection_callbacks::on_connection_id_status(
+            ngtcp2_conn* /* _conn */,
+            ngtcp2_connection_id_status_type type,
+            uint64_t /* seq */,
+            const ngtcp2_cid* cid,
+            const uint8_t* token,
+            void* user_data)
+    {
+        auto* conn = static_cast<Connection*>(user_data);
+        assert(conn->stateless_reset_enabled());
```

Comment:
Why must this be true?

Reply:
I see that this callback is only set up when stateless reset is enabled, but what I don't understand is why that is the case.  There is some commented out code in the removed part of the diff to disassociate the cid when this is called, and it's not clear to me why that isn't the right thing to do when the other side deactivates a dcid.